File: adapter-vsto/src/MathCursor/Host/Detection/ZoneRefiner.cs
using System;
using System.Collections.Generic;
using MathCursor.Detection;

namespace MathCursor.Host.Detection
{
    /// <summary>
    /// Helpers purs pour ajuster les <see cref="DetectedZone"/> produites
    /// par le NER avant qu'elles soient passées à l'engine.
    ///
    /// <para>Bounded context "raffinage de zone détectée" (P2.14 du refactor
    /// archi). Toutes les méthodes sont statiques et purement fonctionnelles —
    /// pas de Word, pas de state. Testables hors Word.</para>
    /// </summary>
    internal static class ZoneRefiner
    {
        /// <summary>
        /// Mots dont la zone détectée juste à droite doit être étendue
        /// rétroactivement (ex. « limite x→0 f(x) » → la zone absorbe
        /// « limite »). Table portée de l'ex data/locale/fr.yml
        /// `math_prefix_keywords:` (Phase 2 beta-clean : le YAML appartenait
        /// à l'ancien moteur).
        /// </summary>
        public static readonly ISet<string> DefaultMathPrefixKeywords =
            new HashSet<string>(StringComparer.OrdinalIgnoreCase)
            {
                "lim", "limite", "lmt",
                "sqrt", "rac", "racine", "racn", "root",
                // iint/iiint : absents du corpus NER — le modèle détecte la
                // queue (« f x y ») sans le mot-clé, l'extension arrière le
                // récupère (retour user 2026-06-11).
                "int", "iint", "iiint", "integrale", "integ", "integral",
                "sum", "somme", "som", "prod", "produit",
                "forall", "qq", "qqe", "pourtout",
                "exists", "existe", "ilexiste",
                "vec", "vect", "vecteur",
                // Sync vocabulaire moteur 2026-06-10. La table est insensible
                // à la casse : « Norm u » (autocapitalisé par Word — cellules
                // du tuto = débuts de phrase) marche dès que « norm » est là.
                "norm", "abs", "module", "conj",
                "cos", "sin", "tan", "ln", "log", "exp",
                "angle", "pgcd", "ppcm", "binom",
                "floor", "ceil",
                // lettres grecques (« pi x », « Delta = … » : la popup ne se
                // déclenchait pas, retour tuto 2026-06-10)
                "alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta",
                "theta", "iota", "kappa", "lambda", "mu", "nu", "xi",
                "omicron", "pi", "rho", "sigma", "tau", "upsilon", "phi",
                "chi", "psi", "omega",
            };

        /// <summary>
        /// Si le caret est juste après la zone avec UNIQUEMENT du whitespace
        /// entre, on pousse l'end de la zone jusqu'au caret (l'user a tapé
        /// un espace pour continuer la formule).
        /// </summary>
        public static DetectedZone TryExtendForwardWhitespace(string paragraph, DetectedZone zone, int caret)
        {
            if (zone == null || string.IsNullOrEmpty(paragraph)) return zone;
            if (caret <= zone.End) return zone;
            int gap = caret - zone.End;
            if (gap > 5) return zone;
            for (int i = zone.End; i < caret && i < paragraph.Length; i++)
                if (!char.IsWhiteSpace(paragraph[i])) return zone;
            int newEnd = Math.Min(caret, paragraph.Length);
            string newText = paragraph.Substring(zone.Start, newEnd - zone.Start);
            return new DetectedZone(zone.Start, newEnd, newText, zone.Confidence);
        }

        /// <summary>
        /// Si le mot juste avant la zone est un math prefix keyword
        /// (cf. <see cref="DefaultMathPrefixKeywords"/>), l'inclut dans la
        /// zone (ex: "limite" + "x→0 f(x)" → 1 zone unique).
        /// </summary>
        public static DetectedZone ExtendBackwardWithKeyword(string paragraph,
            DetectedZone zone,
            ISet<string> mathPrefixKeywords)
        {
            if (string.IsNullOrEmpty(paragraph) || zone == null) return zone;

            int i = zone.Start;
            while (i > 0 && char.IsWhiteSpace(paragraph[i - 1])) i--;
            int wordEnd = i;
            while (i > 0 && char.IsLetter(paragraph[i - 1])) i--;
            int wordStart = i;
            if (wordEnd <= wordStart) return zone;

            string prevWord = paragraph.Substring(wordStart, wordEnd - wordStart);
            if (mathPrefixKeywords == null || !mathPrefixKeywords.Contains(prevWord)) return zone;

            int newEnd = zone.End;
            int newStart = wordStart;
            if (newStart >= 0 && newEnd <= paragraph.Length && newEnd > newStart)
            {
                string newText = paragraph.Substring(newStart, newEnd - newStart);
                return new DetectedZone(newStart, newEnd, newText, zone.Confidence);
            }
            return zone;
        }

        /// <summary>
        /// Jette les zones NER qui chevauchent une région OMath : ces zones
        /// sont déjà converties, pas besoin de les re-proposer.
        /// </summary>
        public static IReadOnlyList<DetectedZone> FilterOutOMathOverlap(
            IReadOnlyList<DetectedZone> zones, IReadOnlyList<(int start, int end)> regions)
        {
            if (zones == null || zones.Count == 0 || regions == null || regions.Count == 0)
                return zones ?? Array.Empty<DetectedZone>();
            var kept = new List<DetectedZone>(zones.Count);
            foreach (var z in zones)
            {
                bool overlaps = false;
                foreach (var (s, e) in regions)
                {
                    if (z.End > s && z.Start < e) { overlaps = true; break; }
                }
                if (!overlaps) kept.Add(z);
            }
            return kept;
        }

        /// <summary>
        /// Choisit la zone la plus proche du caret. Distance = 0 si dedans ou
        /// collée au bord, sinon nombre de chars entre caret et bord le plus
        /// proche de la zone.
        /// </summary>
        public static DetectedZone PickNearestZone(IReadOnlyList<DetectedZone> zones, int caret, out int bestDist)
        {
            DetectedZone best = null;
            bestDist = int.MaxValue;
            foreach (var z in zones)
            {
                int dist;
                if (caret >= z.Start && caret <= z.End) dist = 0;
                else if (caret < z.Start) dist = z.Start - caret;
                else dist = caret - z.End;
                if (dist < bestDist)
                {
                    bestDist = dist;
                    best = z;
                }
            }
            return best;
        }
    }
}

File: rust/mc-ner/src/zone_refiner.rs
// MathCursor — capture d'intention mathématique depuis une saisie clavier linéaire.
// Copyright (C) 2026  Côme Percin
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

//! Raffinage des zones NER — port pur de ZoneRefiner.cs.
//!
//! Opère sur des UNITÉS UTF-16 (`&[u16]`) + offsets UTF-16, pour coïncider avec
//! les positions de l'hôte (VSCode Range / C# char). Classification de caractère
//! sur l'unité décodée (BMP) — les demi-surrogates (astraux, ~absents en maths)
//! ne sont ni blancs ni lettres, comportement sûr.

use crate::DetectedZone;

fn is_ws(u: u16) -> bool {
    char::from_u32(u as u32).map(|c| c.is_whitespace()).unwrap_or(false)
}
fn is_letter(u: u16) -> bool {
    char::from_u32(u as u32).map(|c| c.is_alphabetic()).unwrap_or(false)
}

/// Zone la plus proche du caret (distance 0 si dedans/au bord).
pub fn pick_nearest_zone(zones: &[DetectedZone], caret: usize) -> Option<DetectedZone> {
    let mut best: Option<&DetectedZone> = None;
    let mut best_dist = usize::MAX;
    for z in zones {
        let dist = if caret >= z.start && caret <= z.end {
            0
        } else if caret < z.start {
            z.start - caret
        } else {
            caret - z.end
        };
        if dist < best_dist {
            best_dist = dist;
            best = Some(z);
        }
    }
    best.cloned()
}

fn is_ws_gap(units: &[u16], from: usize, to: usize, max_gap: usize) -> bool {
    if to < from || to - from > max_gap {
        return false;
    }
    for i in from..to {
        if i < units.len() && !is_ws(units[i]) {
            return false;
        }
    }
    true
}

/// Fusionne les zones adjacentes séparées uniquement par des blancs (le NER
/// fragmente parfois une formule). Retourne le pivot tel quel si rien à fusionner.
pub fn merge_whitespace_adjacent(
    zones: &[DetectedZone],
    units: &[u16],
    pivot: DetectedZone,
    max_gap: usize,
) -> DetectedZone {
    let mut start = pivot.start;
    let mut end = pivot.end;
    let mut grew = true;
    while grew {
        grew = false;
        for z in zones {
            if z.end <= start && is_ws_gap(units, z.end, start, max_gap) && z.start < start {
                start = z.start;
                grew = true;
            } else if z.start >= end && is_ws_gap(units, end, z.start, max_gap) && z.end > end {
                end = z.end;
                grew = true;
            }
        }
    }
    if start == pivot.start && end == pivot.end {
        return pivot;
    }
    DetectedZone { start, end: end.min(units.len()), confidence: pivot.confidence }
}

/// Si le caret suit la zone avec UNIQUEMENT du blanc entre (≤ 5), pousse la fin
/// jusqu'au caret (l'user a tapé un espace pour continuer la formule).
pub fn try_extend_forward_whitespace(units: &[u16], zone: DetectedZone, caret: usize) -> DetectedZone {
    if caret <= zone.end {
        return zone;
    }
    if caret - zone.end > 5 {
        return zone;
    }
    for i in zone.end..caret {
        if i < units.len() && !is_ws(units[i]) {
            return zone;
        }
    }
    DetectedZone { start: zone.start, end: caret.min(units.len()), confidence: zone.confidence }
}

/// Si le mot juste avant la zone est un mot-clé math (lim/racine/somme…), l'inclut.
pub fn extend_backward_with_keyword(units: &[u16], zone: DetectedZone) -> DetectedZone {
    let mut i = zone.start;
    while i > 0 && is_ws(units[i - 1]) {
        i -= 1;
    }
    let word_end = i;
    while i > 0 && is_letter(units[i - 1]) {
        i -= 1;
    }
    let word_start = i;
    if word_end <= word_start {
        return zone;
    }
    let word: String = char::decode_utf16(units[word_start..word_end].iter().copied())
        .map(|r| r.unwrap_or('\u{FFFD}'))
        .collect();
    if !MATH_PREFIX_KEYWORDS.contains(&word.to_lowercase().as_str()) {
        return zone;
    }
    DetectedZone { start: word_start, end: zone.end, confidence: zone.confidence }
}

/// Mots-clés dont la zone détectée juste à droite absorbe le mot (table portée
/// de ZoneRefiner.DefaultMathPrefixKeywords, comparaison insensible à la casse).
const MATH_PREFIX_KEYWORDS: &[&str] = &[
    "lim", "limite", "lmt",
    "sqrt", "rac", "racine", "racn", "root",
    "int", "iint", "iiint", "integrale", "integ", "integral",
    "sum", "somme", "som", "prod", "produit",
    "forall", "qq", "qqe", "pourtout",
    "exists", "existe", "ilexiste",
    "vec", "vect", "vecteur",
    "norm", "abs", "module", "conj",
    "cos", "sin", "tan", "ln", "log", "exp",
    "angle", "pgcd", "ppcm", "binom",
    "floor", "ceil",
    "alpha", "beta", "gamma", "delta", "epsilon", "zeta", "eta",
    "theta", "iota", "kappa", "lambda", "mu", "nu", "xi",
    "omicron", "pi", "rho", "sigma", "tau", "upsilon", "phi",
    "chi", "psi", "omega",
];
